src/components/ha-selector/ha-selector-selector.ts

```diff
@@ -0,0 +1,286 @@
+
+  @property() public label?: string;
+
+  @property() public helper?: string;
+
+  @property({ type: Boolean, reflect: true }) public disabled = false;
+
+  @property({ type: Boolean, reflect: true }) public required = true;
+
+  private _yamlMode = false;
+
+  protected shouldUpdate(changedProps: PropertyValues) {
+    if (changedProps.size === 1 && changedProps.has("hass")) {
+      return false;
+    }
+    return true;
+  }
+
+  private _schema = memoizeOne(
+    (choice: string, localize: LocalizeFunc) =>
+      [
+        {
+          name: "type",
+          selector: {
+            select: {
+              mode: "dropdown",
+              required: true,
+              options: Object.keys(SELECTOR_SCHEMAS)
+                .concat("manual")
+                .map((key) => ({
+                  label:
+                    localize(
+                      `ui.components.selectors.selector.types.${key}` as LocalizeKeys
+                    ) || key,
+                  value: key,
+                })),
+            },
+          },
+        },
+        ...(choice === "manual"
+          ? ([
+              {
+                name: "manual",
+                selector: { object: {} },
+              },
+            ] as const)
+          : []),
+        ...(SELECTOR_SCHEMAS[choice]
+          ? SELECTOR_SCHEMAS[choice].length > 1
+            ? [
+                {
+                  name: "",
+                  type: "expandable",
+                  title: localize("ui.components.selectors.selector.options"),
+                  schema: SELECTOR_SCHEMAS[choice],
+                },
+              ]
+            : SELECTOR_SCHEMAS[choice]
+          : []),
+      ] as const
+  );
+
+  protected render() {
+    let data;
+    let type;
+    if (this._yamlMode) {
+      type = "manual";
+      data = { type, manual: this.value };
+    } else {
+      type = Object.keys(this.value)[0];
+      const value0 = Object.values(this.value)[0];
+      data = {
+        type,
+        ...(typeof value0 === "object" ? value0 : []),
+      };
+    }
+
+    const schema = this._schema(type, this.hass.localize);
+
+    return html`<ha-card>
+      <div class="card-content">
+        <p>${this.label ? this.label : ""}</p>
+        <ha-form
+          .hass=${this.hass}
+          .data=${data}
+          .schema=${schema}
+          .computeLabel=${this._computeLabelCallback}
+          @value-changed=${this._valueChanged}
+        ></ha-form></div
+    ></ha-card>`;
+  }
+
+  private _valueChanged(ev: CustomEvent) {
+    ev.stopPropagation();
+    const value = ev.detail.value;
+
+    const type = value.type;
+    if (!type || typeof value !== "object" || Object.keys(value).length === 0) {
+      // not sure how this happens, but reject it
+      return;
+    }
+
+    const oldType = Object.keys(this.value)[0];
+    if (type === "manual" && !this._yamlMode) {
+      this._yamlMode = true;
+      this.requestUpdate();
```

Comment:
Why not make `_yamlMode` a state instead of manually calling `requestUpdate`?

Reply:
So I looked at this one more time, and tried setting `_yamlMode` to a `state`. 

For some reason I can't figure out when I do this, when I change from the manual mode to one of the other dropdown items (therefore leaving yamlMode), the ha-selector-select is resetting itself to -1. I looked at both the data and schema being passed to `ha-form` during render, but it looks correct to me, so I couldn't figure out why it is doing this. 

The only difference between this and the working version is that in the working version after changing the selector we do a single render pass: we fire valueChanged, and then the component rerenders when `value` property changes to reflect that.

In the non-working version, we rerender the component twice, the first time due to the state variable, and then the second time due to the change in `value`. But seems like in this second render the select dropdown wants to reset itself for some reason. 

I feel like I come across these similar kind of problems with the `ha-select` component in lots of different contexts, it either seems to have some bugs I can't figure out, or there's some subtlety about using it that I'm missing. 

Example screen recording:

![selector-selector-2](https://github.com/home-assistant/frontend/assets/32912880/8f0c16da-7c20-40f4-83f1-a372a62be41f)

Reply:
The select probably fires a changed event when it is rendered the first time without the updated value?

Reply:
There is something weird going on in ha-select, but I failed to find a solution... going to merge this